prow/cmd/phaino/main.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type options struct {
+	keepGoing    bool
+	printCmd     bool
+	priv         bool
+	timeout      time.Duration
+	totalTimeout time.Duration
+	grace        time.Duration
+
+	jobs []string
+}
+
+func gatherOptions() options {
+	var o options
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.BoolVar(&o.keepGoing, "keep-going", false, "Continue running jobs after one fails if set")
+	fs.BoolVar(&o.printCmd, "print", false, "Just print the command it would run")
+	fs.BoolVar(&o.priv, "privileged", false, "Allow privileged local runs")
+	fs.DurationVar(&o.timeout, "timeout", 10*time.Minute, "Maximum duration for each job (0 for unlimited)")
+	fs.DurationVar(&o.totalTimeout, "total-timeout", 0, "Maximum duration for all jobs (0 for unlimited)")
+	fs.DurationVar(&o.grace, "grace", 10*time.Second, "Terminate timed out jobs after this grace period (1s minimum)")
+	fs.Parse(os.Args[1:])
+	o.jobs = fs.Args()
+	return o
+}
+
+func readPJ(reader io.Reader) (*prowapi.ProwJob, error) {
+	var pj prowapi.ProwJob
+	buf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("read: %v", err)
+	}
+	if err := yaml.Unmarshal(buf, &pj); err != nil {
+		return nil, fmt.Errorf("unmarshal: %v", err)
+	}
+	if pj.Kind != "ProwJob" {
+		return nil, fmt.Errorf("bad kind: %q", pj.Kind)
+	}
```

Comment:
We also need to require `agent: kubernetes` and that the podspec is non-nil.

Once we have a ProwJob admission controller we can share the ProwJob validation logic here.

Reply:
Done